### ISSUE TITLE
/newspapers endpoint is slow even with caching enabled

### ISSUE THREAD

**Issue by theorm** [impresso/impresso-middle-layer#459]
It does not get better after multiple tries.

**Comment by theorm** [comment]
After looking into it I can confirm that calls to newspaper service are cached - every subsequent call should be fast.
I found DB queries to be very slow:

```
SELECT count(DISTINCT(`newspaper`.`id`)) AS `count` FROM `newspapers` AS `newspaper` LEFT OUTER JOIN `np_timespan_v` AS `stats` ON `newspaper`.`id` = `stats`.`newspaper_id` WHERE (`newspaper`.`title` LIKE '%b%' OR `newspaper`.`id` LIKE '%b%');
+3s
SELECT `newspaper`.`id` AS `uid`, `newspaper`.`title` AS `name`, `newspaper`.`start_year` AS `startYear`, `newspaper`.`end_year` AS `endYear`, `stats`.`start` AS `stats.startYear`, `stats`.`end` AS `stats.endYear`, `stats`.`number_issues` AS `stats.countIssues`, `stats`.`newspaper_id` AS `stats.newspaper_id` FROM `newspapers` AS `newspaper` LEFT OUTER JOIN `np_timespan_v` AS `stats` ON `newspaper`.`id` = `stats`.`newspaper_id` WHERE (`newspaper`.`title` LIKE '%b%' OR `newspaper`.`id` LIKE '%b%') ORDER BY `uid` DESC LIMIT 0, 10; 
+247ms
SELECT count(DISTINCT(`newspaper`.`id`)) AS `count` FROM `newspapers` AS `newspaper` LEFT OUTER JOIN `np_timespan_v` AS `stats` ON `newspaper`.`id` = `stats`.`newspaper_id` WHERE (`newspaper`.`title` LIKE '%b%' OR `newspaper`.`id` LIKE '%b%'); 
+9s
SELECT `newspaper`.`id` AS `uid`, `newspaper`.`title` AS `name`, `newspaper`.`start_year` AS `startYear`, `newspaper`.`end_year` AS `endYear`, `stats`.`start` AS `stats.startYear`, `stats`.`end` AS `stats.endYear`, `stats`.`number_issues` AS `stats.countIssues`, `stats`.`newspaper_id` AS `stats.newspaper_id` FROM `newspapers` AS `newspaper` LEFT OUTER JOIN `np_timespan_v` AS `stats` ON `newspaper`.`id` = `stats`.`newspaper_id` WHERE (`newspaper`.`title` LIKE '%b%' OR `newspaper`.`id` LIKE '%b%') ORDER BY `uid` DESC LIMIT 0, 10; 
+554ms

info: ⏱: "newspapers.find.db.newspapers" took 18863.05 ms.
```

It looks like it's all because of the `np_timespan_v` view. This view returns only about 100 results but it uses the `issues` table which is very big. We need to see if we can optimise the indexes used to build the view. If it's not possible we can select all newspapers in IML, cache them and use the cached version to filter down further.

**Comment by e-maud** [comment]
It seems that this query:



```sql
SELECT
	`newspaper`.`id` AS `uid`,
	`newspaper`.`title` AS `name`,
	`newspaper`.`start_year` AS `startYear`,
	`newspaper`.`end_year` AS `endYear`,
	`stats`.`start` AS `stats.startYear`,
	`stats`.`end` AS `stats.endYear`,
	`stats`.`number_issues` AS `stats.countIssues`,
	`stats`.`newspaper_id` AS `stats.newspaper_id` 
FROM
	`newspapers` AS `newspaper`
	LEFT OUTER JOIN `np_timespan_v` AS `stats` ON `newspaper`.`id` = `stats`.`newspaper_id` 
WHERE
	( `newspaper`.`title` LIKE '%b%' OR `newspaper`.`id` LIKE '%b%' ) 
ORDER BY
	`uid` DESC 
	LIMIT 0,
	10;
```
+247ms

differs from the contents of `np_timespan_v` just by also having the newspaper title - shall we simply add the title to the  view and then it is available?

**Comment by theorm** [comment]
I think we should stop using the view in ad-hoc queries. I started digging into newspapers code in IML and found that newspapers have been cached already and it's used by some code, but not the one that serves the newspaper page. This query will be slow because it joins three tables: newspapers, issues and pages. We discussed it with Daniele and the best way forward is to cache the newspapers periodically and use the cached version to filter the list of newspapers in memory.
